### PR TITLE
Add Conformer RNN-T LibriSpeech training recipe

### PR DESCRIPTION
**Pull request by hwangjeff**
Adds Conformer RNN-T LibriSpeech training recipe to examples directory.

Produces 30M-parameter model that achieves the following WER:

|                     |          WER |
|:-------------------:|-------------:|
| test-clean          |       0.0310 |
| test-other          |       0.0805 |
| dev-clean           |       0.0314 |
| dev-other           |       0.0827 |

### CODE REVIEW

File: examples/asr/librispeech_conformer_rnnt/lightning.py [hunk 1]
```diff
@@ -0,0 +1,350 @@
...
+            current_batch = [idx]
+            current_token_count = target_length
+        else:
+            current_batch.append(idx)
+            current_token_count += target_length
+
+    if current_batch:
+        batches.append(current_batch)
+
+    return batches
```
[the remark this answers is not in the document]
Comment (hwangjeff) [reply examples/asr/librispeech_conformer_rnnt/lightning.py:51]: yeah we can figure out how to consolidate later. i don't want the two directories to be too entangled

File: examples/asr/librispeech_conformer_rnnt/lightning.py [hunk 2]
```diff
@@ -0,0 +1,350 @@
...
+    def training_step(self, batch: Batch, batch_idx):
+        opt = self.optimizers()
+        opt.zero_grad()
+        loss = self._step(batch, batch_idx, "train")
+        batch_size = batch.features.size(0)
+        batch_sizes = self.all_gather(batch_size)
+        self.log("Gathered batch size", batch_sizes.sum(), on_step=True, on_epoch=True)
+        loss *= batch_sizes.size(0) / batch_sizes.sum()  # world size / batch size
+        self.manual_backward(loss)
+        torch.nn.utils.clip_grad_norm_(self.model.parameters(), 10.0)
```
[the remark this answers is not in the document]
Comment (hwangjeff) [reply examples/asr/librispeech_conformer_rnnt/lightning.py:283]: actually, i don't currently surface this as an option, and i don't foresee users being interested in tuning this at this point, so i'm fine with keeping it as is

File: examples/asr/librispeech_conformer_rnnt/lightning.py [hunk 3]
```diff
@@ -0,0 +1,350 @@
...
+
+        self.mean = torch.tensor(blob["mean"])
+        self.invstddev = torch.tensor(blob["invstddev"])
+
+    def forward(self, input):
+        return (input - self.mean) * self.invstddev
+
+
+class WarmupLR(torch.optim.lr_scheduler._LRScheduler):
+    def __init__(self, optimizer, warmup_updates, force_anneal_epoch, anneal_factor, last_epoch=-1, verbose=False):
```
[the remark this answers is not in the document]
Comment (hwangjeff) [reply examples/asr/librispeech_conformer_rnnt/lightning.py]: i changed `warmup_updates` to `warmup_steps` and `force_anneal_epoch` to `force_anneal_step` to reflect that `WarmupLR` is agnostic to the frequency of scheduler steps

File: examples/asr/librispeech_conformer_rnnt/train.py [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+        sp_model_path=str(args.sp_model_path),
+        global_stats_path=str(args.global_stats_path),
+    )
+    trainer.fit(model)
+
+
+def cli_main():
+    parser = ArgumentParser()
+    parser.add_argument(
+        "--exp_dir",
```
[the remark this answers is not in the document]
Comment (hwangjeff) [reply examples/asr/librispeech_conformer_rnnt/train.py]: done

**Review by xiaohui-zhang — APPROVED** [review]

File: examples/asr/librispeech_conformer_rnnt/lightning.py [hunk 5]
```diff
@@ -0,0 +1,390 @@
...
+    nbest_batch = list(zip(hypos_str, hypos_score, hypos_ali, hypos_ids))
+
+    return nbest_batch
+
+
+class RNNTModule(LightningModule):
+    def __init__(
+        self,
+        *,
+        librispeech_path: str,
```
[the remark this answers is not in the document]
Comment (hwangjeff) [reply examples/asr/librispeech_conformer_rnnt/lightning.py:192]: for now, the module instantiates the librispeech dataset internally, so specifying "librispeech" is clearer

File: examples/asr/librispeech_conformer_rnnt/lightning.py [hunk 6]
```diff
@@ -0,0 +1,390 @@
...
+        hypotheses = decoder(batch.features.to(self.device), batch.feature_lengths.to(self.device), 20)
+        return post_process_hypos(hypotheses, self.sp_model)[0][0]
+
+    def training_step(self, batch: Batch, batch_idx):
+        """Custom training step.
+
+        By default, DDP does the following on each train step:
+        - For each GPU, compute loss and gradient on shard of training data.
+        - Sync and average gradients across all GPUs. The final gradient
+          is (sum of gradients across all GPUs) / N, where N is the world
```
[the remark this answers is not in the document]
Comment (hwangjeff) [reply examples/asr/librispeech_conformer_rnnt/lightning.py:300]: it should be "world" — this appears to be standard terminology in distributed training, e.g. https://pytorch.org/docs/stable/distributed.html#torch.distributed.get_world_size

File: examples/asr/librispeech_conformer_rnnt/train.py [hunk 7]
```diff
@@ -0,0 +1,101 @@
...
+        help="Path to SentencePiece model.",
+    )
+    parser.add_argument(
+        "--num-nodes",
+        default=4,
+        type=int,
+        help="Number of nodes to use for training. (Default: 4)",
+    )
+    parser.add_argument(
+        "--gpus",
```
[the remark this answers is not in the document]
Comment (hwangjeff) [reply examples/asr/librispeech_conformer_rnnt/train.py:84]: changed `num-nodes` to `nodes`

File: examples/asr/librispeech_conformer_rnnt/lightning.py [hunk 8]
```diff
@@ -0,0 +1,390 @@
...
+    hypos_ali = [h.alignment[1:] for h in hypos]
+    hypos_ids = [h.tokens[1:] for h in hypos]
+    hypos_score = [[math.exp(h.score)] for h in hypos]
+
+    nbest_batch = list(zip(hypos_str, hypos_score, hypos_ali, hypos_ids))
+
+    return nbest_batch
+
+
+class RNNTModule(LightningModule):
```
[the remark this answers is not in the document]
Comment (hwangjeff) [reply examples/asr/librispeech_conformer_rnnt/lightning.py]: changed

File: examples/asr/librispeech_conformer_rnnt/README.md [hunk 9]
```diff
@@ -0,0 +1,32 @@
...
+
+## Usage
+
+### Training
+
+[`train.py`](./train.py) trains an Conformer RNN-T model (30.2M parameters, 121MB) on LibriSpeech using PyTorch Lightning. Note that the script expects users to have access to GPU nodes for training and provide paths to the full LibriSpeech dataset and the SentencePiece model to be used to encode targets.
+
+Sample SLURM command:
+```
+srun --cpus-per-task=12 --gpus-per-node=8 -N 4 --ntasks-per-node=8 python train.py --exp_dir ./experiments --librispeech_path ./librispeech/ --global_stats_path ./global_stats.json --sp_model_path ./spm_unigram_1023.model --epochs 160
```
[the remark this answers is not in the document]
Comment (hwangjeff) [reply examples/asr/librispeech_conformer_rnnt/README.md:30]: added above

**Review by mthrok — APPROVED** [review]
Stamp.
